File: app/src/main/java/com/borisruzanov/russianwives/mvp/presenter/SearchPresenter.java
package com.borisruzanov.russianwives.mvp.presenter;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.borisruzanov.russianwives.models.SearchModel;
import com.borisruzanov.russianwives.models.User;
import com.borisruzanov.russianwives.mvp.model.interactor.SearchInteractor;
import com.borisruzanov.russianwives.mvp.view.SearchView;
import com.borisruzanov.russianwives.utils.UsersListCallback;

import java.util.Arrays;
import java.util.List;

/**
 * Created by Михаил on 03.05.2018.
 */

@InjectViewState
public class SearchPresenter extends MvpPresenter<SearchView> {

    private SearchInteractor searchInteractor;

    public SearchPresenter(SearchInteractor searchInteractor) {
        this.searchInteractor = searchInteractor;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        List<SearchModel> searchModels = Arrays.asList(new SearchModel("name", "John"),
                new SearchModel("hobby", "adventures"));
        searchInteractor.searchByListParams(searchModels, new UsersListCallback() {
            @Override
            public void getUsers(List<User> userList) {
                if(userList.isEmpty()) Log.d("Search", "Search list is empty");

                for (User model: userList) {
                    Log.d("Search", model.getName());
                }
            }
        });
    }

    public void getUsers(){
        searchInteractor.getUsers(new UsersListCallback() {
            @Override
            public void getUsers(List<User> userList) {
                if(!userList.isEmpty()) getViewState().showUsers(userList);
                else getViewState().showEmpty(true);
            }
        });
    }

}

File: app/src/main/java/com/borisruzanov/russianwives/mvp/ui/search/SearchPresenter.kt
package com.borisruzanov.russianwives.mvp.ui.search

import android.os.Bundle
import android.util.Log

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.borisruzanov.russianwives.models.FsUser
import com.borisruzanov.russianwives.mvp.model.interactor.search.SearchInteractor
import com.borisruzanov.russianwives.utils.BoolCallback
import com.borisruzanov.russianwives.utils.FirebaseUtils.isUserExist
import com.borisruzanov.russianwives.utils.StringsCallback
import com.borisruzanov.russianwives.utils.UsersListCallback

import java.util.ArrayList
import javax.inject.Inject

@InjectViewState
class SearchPresenter @Inject constructor(private val searchInteractor: SearchInteractor) : MvpPresenter<SearchView>() {
    private val fsUsers = ArrayList<FsUser>()
    private val likedUsers = HashSet<String>()

    fun getUserList(page: Int) {
        searchInteractor.getFilteredUserList(usersListCallback, page)
        Log.d("UsersListDebug", "in getUserList and fsUsers size is ${fsUsers.size}")
    }

    private val usersListCallback = UsersListCallback { userList ->
        if (userList.isNotEmpty()) {
            if (!fsUsers.containsAll(userList)) {
                Log.d("UsersListDebug", "Users list size is ${userList.size}")
                if (fsUsers.isNotEmpty() && fsUsers.last().uid == userList.last().uid){
                    userList.remove(userList.last())
                }
                Log.d("UsersListDebug", "Users list size after removing is  ${userList.size}")
                fsUsers.addAll(userList)
                Log.d("UsersListDebug", "Add users to fsUsers and fsUsers size is " + fsUsers.size)
            }
            viewState.addUsers(userList)
        }
    }

    fun setProgressBar(isLoading: Boolean) {
        viewState.setProgressBar(isLoading)
    }

    fun onUpdate() {
        fsUsers.clear()
        likedUsers.clear()
        viewState.clearUsers()
        getUserList(0)
    }

    fun openSliderWithDefaults() {
        searchInteractor.getDefaultList(callback = StringsCallback { strings ->
            viewState.openSlider(ArrayList<String>(strings)) })
    }

    fun setFriendLiked(position: Int) {
        Log.d("LikeDebug", "Liked Position is $position")
        if (isUserExist()) {
            val friendModel = fsUsers[position]
            if (fsUsers.contains(friendModel)) {
                Log.d("LikeDebug", "Friend name is ${friendModel.name} and fsUsers size is ${fsUsers.size}")
            }
            searchInteractor.isFriendLiked(fsUsers[position].uid, callback = BoolCallback { hasLiked ->
                if (!hasLiked) {
                    searchInteractor.setFriendLiked(fsUsers[position].uid)
                    searchInteractor.addRatingLiked(fsUsers[position].uid)
                    Log.d("LikeDebug", "${fsUsers[position].name} was liked")
                }
                else Log.d("LikeDebug", "${fsUsers[position].name} WAS NOT liked")
            })
        } else viewState.showRegistrationDialog()
    }

    fun openFriend(position: Int, args: Bundle) {
        viewState.openFriend(fsUsers[position].uid, fsUsers[position].name, args)
    }

    fun openChat(position: Int) {
        if (isUserExist()){
            searchInteractor.hasMustInfo(callback = BoolCallback { hasMustInfo ->
                if (hasMustInfo) viewState.openChat(fsUsers[position].uid, fsUsers[position].name, fsUsers[position].image)
                else viewState.showFullProfileDialog()
            })
        }
        else viewState.showRegistrationDialog()
    }

}
